### PR TITLE
Item Expiration

### PR DESCRIPTION
**Pull request by Zintixx**

### CODE REVIEW

File: Maple2.Model/Game/Item/Item.cs [hunk 1]
```diff
@@ -173,6 +176,28 @@ public class Item : IByteSerializable, IByteDeserializable {
...
+
+        if (Metadata.Life.ExpirationPeriod != default) {
+            expirationTime = (long) (Metadata.Life.ExpirationPeriod.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds;
+        } else if (Metadata.Life.UsePeriod > 0) {
+            expirationTime = (long) (DateTime.Now.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds + Metadata.Life.UsePeriod;
+        } else if (Metadata.Life.ExpirationType != ItemExpirationType.None) {
+            expirationTime = Metadata.Life.ExpirationType switch
+            {
+                ItemExpirationType.Months => Metadata.Life.ExpirationTypeDuration * 2628000 + (long) (DateTime.Now.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds,
+                ItemExpirationType.Weeks => Metadata.Life.ExpirationTypeDuration * 604800 + (long) (DateTime.Now.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds,
```
Comment (kOchirasu) [inline Maple2.Model/Game/Item/Item.cs:193]: `TimeSpan.FromDays(...).Seconds`

File: Maple2.Model/Game/Item/Item.cs [hunk 2]
```diff
@@ -173,6 +176,28 @@ public class Item : IByteSerializable, IByteDeserializable {
...
+        }
+
+        if (Metadata.Life.ExpirationPeriod != default) {
+            expirationTime = (long) (Metadata.Life.ExpirationPeriod.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds;
+        } else if (Metadata.Life.UsePeriod > 0) {
+            expirationTime = (long) (DateTime.Now.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds + Metadata.Life.UsePeriod;
+        } else if (Metadata.Life.ExpirationType != ItemExpirationType.None) {
+            expirationTime = Metadata.Life.ExpirationType switch
+            {
+                ItemExpirationType.Months => Metadata.Life.ExpirationTypeDuration * 2628000 + (long) (DateTime.Now.ToUniversalTime() - DateTime.UnixEpoch).TotalSeconds,
```
Comment (kOchirasu) [inline Maple2.Model/Game/Item/Item.cs:192]: How was this seconds/month (2628000) value determined?

File: Maple2.File.Ingest/Mapper/ItemMapper.cs [hunk 3]
```diff
@@ -82,6 +88,12 @@ public class ItemMapper : TypeMapper<ItemMetadata> {
                     .Select(slot => Enum.Parse<EquipSlot>(slot.name, true))
                     .ToArray(),
                 Mesh: data.ucc.mesh,
+                Life: new ItemMetadataLife(
+                    UsePeriod: data.life.usePeriod,
+                    ExpirationPeriod: expirationTime,
+                    ExpirationType: (ItemExpirationType) data.life.expirationType,
```
Comment (kOchirasu) [inline Maple2.File.Ingest/Mapper/ItemMapper.cs:94]: I feel like we can just do the conversion here at parse time and just store duration in seconds rather than calculating it in `Item.cs`

Basically just store `ExpirationDuration` rather than `ExpirationType`+`ExpirationTypeDuration`

File: Maple2.Model/Metadata/ItemMetadata.cs [hunk 4]
```diff
@@ -15,6 +17,12 @@ public record ItemMetadata(
     ItemMetadataMusic? Music,
     ItemMetadataHousing? Housing) : ISearchResult;
 
+public record ItemMetadataLife(
+    int UsePeriod,
+    DateTime ExpirationPeriod,
```
Comment (kOchirasu) [inline Maple2.Model/Metadata/ItemMetadata.cs:22]: It's probably better not to store as `DateTime` for metadata entries since we don't have a conversion layer. (Metadata types are directly used by both DB and Server due to immutability). MySQL `DateTime` is stored using local system timezone which can technically vary, while I've standardized the game layer to use UTC (even though it's technically wrong, at least it's consistent)

tl;dr Let's use long UTC epoch timestamp for now.

File: Maple2.File.Ingest/Mapper/ItemMapper.cs [hunk 5]
```diff
@@ -82,6 +101,10 @@ public class ItemMapper : TypeMapper<ItemMetadata> {
                     .Select(slot => Enum.Parse<EquipSlot>(slot.name, true))
                     .ToArray(),
                 Mesh: data.ucc.mesh,
+                Life: new ItemMetadataLife(
```
Comment (kOchirasu) [inline Maple2.File.Ingest/Mapper/ItemMapper.cs:104]: `Life` is marked nullable, but it's never null here
